packages/itwin/tree-widget/src/test/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.test.ts

```diff
@@ -0,0 +1,2266 @@
+            const modelId = "0x1";
+            const node = createModelHierarchyNode(modelId);
+            const viewport = createFakeSinonViewport();
+            const { handler } = createHandler({ viewport });
+            await handler.changeVisibility(node, true);
+            expect(viewport.addViewedModels).to.be.calledOnceWith(modelId);
+          });
+
+          testAlwaysAndNeverDrawnChildrenAccess(true);
+
+          it(`removes per model category overrides`, async () => {
+            const modelId = "0x1";
+            const categoryIds = ["0x2", "0x3", "0x4"];
+            const node = createModelHierarchyNode(modelId);
+            const viewport = createFakeSinonViewport();
+            const { handler } = createHandler({
+              viewport,
+              idsCache: createFakeIdsCache({
+                modelCategories: new Map([[modelId, categoryIds]]),
+              }),
+            });
+            await handler.changeVisibility(node, true);
+
+            expect(viewport.perModelCategoryVisibility.clearOverrides).to.be.calledWith(modelId);
+          });
+        });
+
+        describe("off", () => {
+          it("marks itself hidden", async () => {
+            const modelId = "0x1";
+            const node = createModelHierarchyNode(modelId);
+            const viewport = createFakeSinonViewport();
+            const { handler } = createHandler({ viewport });
+            await handler.changeVisibility(node, false);
+            expect(viewport.changeModelDisplay).to.be.calledOnceWith(modelId, false);
+          });
+
+          testAlwaysAndNeverDrawnChildrenAccess(false);
+
+          it(`removes per model category overrides`, async () => {
+            const modelId = "0x1";
+            const categoryIds = ["0x2", "0x3", "0x4"];
+            const node = createModelHierarchyNode(modelId);
+            const viewport = createFakeSinonViewport();
+            const { handler } = createHandler({
+              viewport,
+              idsCache: createFakeIdsCache({
+                modelCategories: new Map([[modelId, categoryIds]]),
+              }),
+            });
+            await handler.changeVisibility(node, false);
+
+            expect(viewport.perModelCategoryVisibility.clearOverrides).to.be.calledWith(modelId);
+          });
+        });
+      });
+
+      describe("category", () => {
+        it("can be overridden", async () => {
+          const overrides: ModelsTreeVisibilityHandlerProps["overrides"] = {
+            changeCategoryState: sinon.fake.resolves(undefined),
+          };
+          const handler = createModelsTreeVisibilityHandler({
+            viewport: createFakeSinonViewport(),
+            overrides,
+          });
+
+          await handler.changeVisibility(createCategoryHierarchyNode("0x1"), true);
+          expect(overrides.changeCategoryState).to.be.called;
+        });
+
+        describe("on", () => {
+          it("removes HIDE override if model is shown", async () => {
```

Comment:
I think we need an additional test case for when:
- the model is displayed,
- there's no per-model category override,
- the category is not in category selector.

Reply:
Added